package/kernel/leds-lp5569/src/leds-lp5569.c

```diff
@@ -0,0 +1,1480 @@
+
+#include "../drivers/leds/leds-lp55xx-common.h"
+
+#define CONFIG_LEDS_LP5569_PREDEFINED_PATTERNS 1
+
+#if defined(CONFIG_LEDS_LP5569_EXTENDED_FW)
+#define LP5569_PROGRAM_PAGES		16
+#endif
+#define LP5569_PROGRAM_LENGTH		32	/* bytes */
+/* Memory is used like this:
+   0x00 engine 1 program
+   0x10 engine 2 program
+   0x20 engine 3 program
+   0x30 engine 1 muxing info
+   0x40 engine 2 muxing info
+   0x50 engine 3 muxing info
+*/
+#define LP5569_MAX_LEDS			9
+
+/* Registers */
+#define LP5569_REG_ENABLE		0x00
+#define LP5569_REG_EXEC_CTRL		0x01
+#define LP5569_REG_OP_MODE		0x02
+#define LP5569_REG_ENABLE_LEDS_MSB	0x04
+#define LP5569_REG_ENABLE_LEDS_LSB	0x05
+#define LP5569_REG_LED_CTRL_BASE	0x07
+#define LP5569_REG_LED_PWM_BASE		0x16
+#define LP5569_REG_LED_CURRENT_BASE	0x22
+#define LP5569_REG_CONFIG		0x2F
+#define LP5569_REG_STATUS		0x3C
+#define LP5569_REG_IO_CONTROL		0x3D
+#define LP5569_REG_RESET		0x3F
+#define LP5569_REG_LED_TEST_CTRL	0x41
+#define LP5569_REG_LED_TEST_ADC		0x42
+#define LP5569_REG_MASTER_FADER_BASE	0x46
+#define LP5569_REG_CH1_PROG_START	0x4B
+#define LP5569_REG_CH2_PROG_START	0x4C
+#define LP5569_REG_CH3_PROG_START	0x4D
+#define LP5569_REG_PROG_PAGE_SEL	0x4F
+#define LP5569_REG_PROG_MEM		0x50
+
+/* Bit description in registers */
+#define LP5569_ENABLE			0x40
+#define LP5569_AUTO_INC			0x40
+#define LP5569_PWR_SAVE			0x20
+#define LP5569_PWM_PWR_SAVE		0x04
+#define LP5569_CP_AUTO			0x18
+#define LP5569_INTERNAL_CLK		0x01
+
+#define LP5569_CLK_PIN_MASK		0x08
+#define LP5569_CLK_OUTPUT		0x08
+
+#define LP5569_EN_LEDTEST		0x80
+#define LP5569_LEDTEST_DONE		0x80
+#define LP5569_RESET			0xFF
+#define LP5569_ADC_SHORTCIRC_LIM	80
+#define LP5569_EXT_CLK_USED		0x08
+#define LP5569_ENG_STATUS_MASK		0x07
+
+#define LP5569_FADER_MAPPING_MASK	0xC0
+#define LP5569_FADER_MAPPING_SHIFT	6
+
+/* Memory Page Selection */
+#define LP5569_PAGE_ENG1		0
+#define LP5569_PAGE_ENG2		1
+#define LP5569_PAGE_ENG3		2
+#define LP5569_PAGE_MUX1		3
+#define LP5569_PAGE_MUX2		4
+#define LP5569_PAGE_MUX3		5
+
+/* Program Memory Operations */
+#define LP5569_MODE_ENG1_M		0xC0	/* Operation Mode Register */
+#define LP5569_MODE_ENG2_M		0x30
+#define LP5569_MODE_ENG3_M		0x0C
+#define LP5569_LOAD_ENG1		0x40
+#define LP5569_LOAD_ENG2		0x10
+#define LP5569_LOAD_ENG3		0x04
+
+#define LP5569_ENG1_IS_LOADING(mode)	\
+	((mode & LP5569_MODE_ENG1_M) == LP5569_LOAD_ENG1)
+#define LP5569_ENG2_IS_LOADING(mode)	\
+	((mode & LP5569_MODE_ENG2_M) == LP5569_LOAD_ENG2)
+#define LP5569_ENG3_IS_LOADING(mode)	\
+	((mode & LP5569_MODE_ENG3_M) == LP5569_LOAD_ENG3)
+
+#define LP5569_EXEC_ENG1_M		0xC0	/* Enable Register */
+#define LP5569_EXEC_ENG2_M		0x30
+#define LP5569_EXEC_ENG3_M		0x0C
+#define LP5569_EXEC_M			0xFC
+#define LP5569_RUN_ENG1			0x80
+#define LP5569_RUN_ENG2			0x20
+#define LP5569_RUN_ENG3			0x08
+
+#define LED_ACTIVE(mux, led)		(!!(mux & (0x0001 << led)))
+
+enum lp5569_chip_id {
+	LP5569,
+};
+
+#if defined(CONFIG_LEDS_LP5569_PREDEFINED_PATTERNS)
```

Comment:
All these predefined program patterns must be removed before submitting to the Linux kernel, everything from line 133 to 377.

Reply:
Thank you for this good hint. I did not notice Zyxel made changes to it.